### PR TITLE
(DOCSP-16043): [iOS] Update documentation and code examples for new property declaration syntax

### PR DESCRIPTION
**Pull request by dacharyc**
### Jira

- https://jira.mongodb.org/browse/DOCSP-16043

### Project Kickoff Doc

- [iOS @Persisted Property Wrapper Docs Project Outline](https://docs.google.com/document/d/1nPr-2qfkJaLnqcdS1EZbo3kF9rHcXd6cwdtLq4bDg3g/edit?usp=sharing)

### Staged Changes (Requires MongoDB Corp SSO)

- [Object Models & Schemas](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/DOCSP-16043/sdk/ios/fundamentals/object-models-and-schemas/)
     - Added new sections and descriptions under the [Property Attributes section](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/DOCSP-16043/sdk/ios/fundamentals/object-models-and-schemas/#property-attributes)
     - Added a note about [changes to Model Inheritance](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/DOCSP-16043/sdk/ios/fundamentals/object-models-and-schemas/#model-inheritance) relative to the new `@Persisted` declaration
- [Supported Property Types](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/DOCSP-16043/sdk/ios/data-types/supported-property-types/)
     - Added the new `@Persisted` property declarations to the `Swift` tab
     - Added a note about performance when setting default values for some types using the new `@Persisted` notation under the `Swift` property declaration cheat sheet tab
     - Added a note about performance implications of the new `@Persisted` syntax under the Unique Identifiers section
     - Added a `Swift pre-10.10.10` tab with the old style property declarations 
- [Map](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/DOCSP-16043/sdk/ios/data-types/map/)
     - Added a note under usage about performance implications when declaring default values
     - Updated code example to show the `@Persisted` syntax
- [MutableSet](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/DOCSP-16043/sdk/ios/data-types/mutablesets/)
     - Added a note under usage about performance implications when declaring default values
     - Updated code example to show the `@Persisted` syntax
- [Define a Realm Object Schema](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/DOCSP-16043/sdk/ios/examples/define-a-realm-object-model/)
     - Update all `Swift` examples for the new `@Persisted` syntax
     - Add a `Swift pre-10.10.10` tab showing the `@objc dynamic` syntax in some places:
          - Declare Properties
          - Specify an Optional/Required Property
          - Specify a Primary Key
          - Index a Property
          - Ignore a Property
          - Declare Enum Properties
          - Define an Inverse Relationship Property
- [Read & Write Data](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/DOCSP-16043/sdk/ios/examples/read-and-write-data/)
     - Update `Swift` examples that show property declarations for the new `@Persisted` syntax
          - The models in [About These Examples](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/DOCSP-16043/sdk/ios/examples/read-and-write-data/#about-these-examples)
          - The object in [Filter on Object ID Properties](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/DOCSP-16043/sdk/ios/examples/read-and-write-data/#filter-on-object-id-properties)
- [Filter Data](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/DOCSP-16043/sdk/ios/examples/filter-data/)
     - Update the  models in the [About the Examples on This Page](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/DOCSP-16043/sdk/ios/examples/filter-data/) to show the new `@Persisted` syntax
- [React to Changes](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/DOCSP-16043/sdk/ios/examples/react-to-changes/)
     - Update the model in the [Register an Object Change Listener](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/DOCSP-16043/sdk/ios/examples/react-to-changes/#register-an-object-change-listener) code example to show the new `@Persisted` syntax
- [Modify an Object Schema](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/DOCSP-16043/sdk/ios/examples/modify-an-object-schema/)
     - Update the model in the [Perform a Schema Migration](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/DOCSP-16043/sdk/ios/examples/modify-an-object-schema/#perform-a-schema-migration) code example to show the new `@Persisted` syntax
- [Quick Start](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/DOCSP-16043/sdk/ios/quick-start/)
     - Update the code examples to show the new `@Persisted` syntax
- [Quick Start with Sync](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/DOCSP-16043/sdk/ios/quick-start-with-sync/)
     - Update the code examples to show the new `@Persisted` syntax

Still to do that is not part of this ticket:
- Update the [SwiftUI & Combine](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/DOCSP-16043/sdk/ios/integrations/swiftui/) Integration Guide to show the new syntax
- Update the [iOS Swift Tutorial](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/DOCSP-16043/tutorial/ios-swift/) to show the new syntax

### Review Guidelines

[REVIEWING.md](https://github.com/mongodb/docs-realm/blob/master/REVIEWING.md)

### CODE REVIEW

File: examples/ios/Examples/ObjectModels.swift [hunk 1]
```diff
@@ -11,19 +13,31 @@ import RealmSwift
...
+    // Required numeric property
+    @Persisted var ageYears = 0
+
+    // Optional numeric property
+    @Persisted var heightCm: Float?
+}
+// :code-block-end:
+
+// :code-block-start: optional-required-properties-objc-dynamic
+class OptionalRequiredPropertyObjcDynamicExample_Person: Object {
```
Comment (cbush) [inline examples/ios/Examples/ObjectModels.swift:40]: [Non-actionable comment] Wow, :replace: again to the rescue 🙌

**Review by tgoyne — APPROVED** [review]
Everything looks reasonable to me.

File: source/sdk/ios/data-types/supported-property-types.txt [hunk 2]
```diff
@@ -148,16 +168,105 @@ Property Cheat Sheet
...
-globally-unique value. You can :ref:`index <ios-index-a-property>` both types,
-and use either as a :ref:`primary key <ios-specify-a-primary-key>`.
+``ObjectId`` is a MongoDB-specific 12-byte unique value. ``UUID`` is a 
+16-byte globally-unique value. You can :ref:`index <ios-index-a-property>` 
+both types, and use either as a :ref:`primary key <ios-specify-a-primary-key>`.
+
+.. note::
+
+   When declaring default values for ``@Persisted`` UUID or ObjectId property 
+   attributes, both of these syntax types is valid:
```
Comment (cbush) [inline source/sdk/ios/data-types/supported-property-types.txt:259]: both... are? Hm...

File: source/sdk/ios/data-types/supported-property-types.txt [hunk 3]
```diff
@@ -148,16 +168,105 @@ Property Cheat Sheet
...
+
+   - ``@Persisted var value: UUID``
+   - ``@Persisted var value = UUID()``
+
+   However, the second will result in poorer performance. 
+   
+   Additionally, ``@Persisted var id: ObjectId`` has equivalent behavior to 
+   ``@objc dynamic var _id = ObjectId.generate()``, but not ``@objc 
+   dynamic var _id = ObjectId()``. The first syntax provides better performance
+   than ``ObjectId.generate()``.
```
Comment (cbush) [inline source/sdk/ios/data-types/supported-property-types.txt:269]: I feel something is getting lost here. It's not entirely clear if you mean that generate() (and the first syntax) actually generates a random ID while the default constructor (initializer, whatever) creates a "blank" ObjectId.

File: source/sdk/ios/fundamentals/object-models-and-schemas.txt [hunk 4]
```diff
@@ -34,6 +34,49 @@ object that can be stored in that {+realm+}.
...
+.. _persisted-property-attributes:
+
+Persisted Property Attributes
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. versionadded:: 10.10.0
+   The ``@Persisted`` declaration style replaces the ``@objc dynamic``, 
+   ``RealmOptional``, and ``RealmProperty`` declaration notations from older 
+   versions of the SDK. For an older version of the SDK, see: 
+   :ref:`Objective-C Dynamic Property Attributes <objc-dynamic-property-attributes>`
```
Comment (cbush) [inline source/sdk/ios/fundamentals/object-models-and-schemas.txt:55]: Missing period:
```suggestion
   :ref:`Objective-C Dynamic Property Attributes <objc-dynamic-property-attributes>`.
```

File: source/sdk/ios/fundamentals/object-models-and-schemas.txt [hunk 5]
```diff
@@ -171,6 +214,13 @@ does not allow you to:
    <realm/realm-cocoa/issues/1109#issuecomment-143834756>` for working
    around these limitations.
 
+.. versionadded:: 10.10.0
+   While you can't mix ``@Persisted`` and ``@objc dynamic`` property declarations
+   within a class, you can mix the notation styles across parent and subclasses. 
+   For example, a parent class could have a ``@Persisted var foo: Int`` property, 
+   and a child of that class could have an ``@objc dynamic var bar = 0`` property,
```
Comment (cbush) [inline source/sdk/ios/fundamentals/object-models-and-schemas.txt:221]: Do you mean base and subclass? Or parent and embedded object classes?

**Review by cbush — APPROVED** [review]
I love the solution here. I think the tabs work pretty well, and keeping the examples by changing the replaced prefix is a stroke of brilliance. I really appreciate all the hard work on this one too, as this had to touch a lot of different areas. Overall, looks great with just a few minor comments/questions.

One other thing is that it seems the table was not fixed by the suggestion of setting the column widths explicitly, so should we raise a DOP issue for that?

File: source/sdk/ios/data-types/supported-property-types.txt (hunk 3 above)
Comment (dacharyc) [reply source/sdk/ios/data-types/supported-property-types.txt:269]: The note from Thomas was that the first two have equivalent behavior, but not the third. And the first provides better performance than the second.

I'm not sure what, if anything, we need to say about `@objc dynamic var _id = ObjectId()` aside from the fact that it is not equivalent to the `@Persisted` syntax that provides the best performance. Maybe I could just lose that part, and leave it `@Persisted` provides equivalent behavior to `ObjectId.generate()`, but better performance?

File: source/sdk/ios/fundamentals/object-models-and-schemas.txt (hunk 5 above)
Comment (dacharyc) [reply source/sdk/ios/fundamentals/object-models-and-schemas.txt:221]: Tried to clarify this a bit - how's:

> While you can't mix ``@Persisted`` and ``@objc dynamic`` property declarations
   within a class, you can mix the notation styles across base and subclasses. 
   For example, a base class could have a ``@Persisted var foo: Int`` property, 
   and a subclass could have an ``@objc dynamic var bar = 0`` property, with 
   both persisted. However, the ``@objc dynamic`` property would be ignored if
   the ``@Persisted`` property was within the same base or subclass.

File: source/sdk/ios/data-types/supported-property-types.txt (hunk 3 above)
Comment (cbush) [reply source/sdk/ios/data-types/supported-property-types.txt]: That's what I'm wondering. @tgoyne can confirm, but ObjectId() creates a _zero-initialized_ ObjectId. So it's more than a performance difference between than and ObjectId.generate(). https://docs.mongodb.com/realm-sdks/swift/latest/Classes/ObjectId.html#/c:@M@RealmSwift@objc(cs)RealmSwiftObjectId(im)init

File: source/sdk/ios/fundamentals/object-models-and-schemas.txt (hunk 5 above)
Comment (cbush) [reply source/sdk/ios/fundamentals/object-models-and-schemas.txt]: Looks good. (Though nitpick on last line: subjunctive, if the property *were* within?)

File: source/sdk/ios/data-types/supported-property-types.txt (hunk 3 above)
Comment (tgoyne) [reply source/sdk/ios/data-types/supported-property-types.txt]: Yeah, `@Persisted var _id = ObjectId()` and `@objc dynamic var _id = ObjectId()` make zero-initialized ObjectIds, while `@Persisted var _id: ObjectId` and `@objc dynamic var _id = ObjectId.generate()` make random ObjectIds (with the latter creating a new random one that's never used every time an object is read from the Realm, while the former only creates them when needed).

Comment (dacharyc) [reply source/sdk/ios/data-types/supported-property-types.txt]: Awesome, thank you both!

**Pull request merged**